From 6522c3e38696821852463d717ceb24738c9b972f Mon Sep 17 00:00:00 2001
From: Steve deRosier <Steve.Derosier@lairdtech.com>
Date: Tue, 10 Dec 2013 20:29:47 -0800
Subject: [PATCH] Fix UART0 and UART3 serial ports for wb45n board

This patch enables the UART0 and UART3 ports:
* It fixes a minor bug in the existing usart3 pinctrl definitions.
* Adds the USART3.
* Adds rts and cts pins for USART0 and USART3.
* Removes unused USART1 and USART2 definitions so they won't make /dev entries
* Disables the use-dma-*x flag to avoid hanging problems.

/dev/ttyS0 is the DEBUG port
/dev/ttyS1 is UART0
/dev/ttyS2 is UART3

Bug: 4524
---
 arch/arm/boot/dts/at91sam9x5.dtsi   |   12 +++++++-----
 arch/arm/boot/dts/at91sam9x5cm.dtsi |   15 +++++++++++++++
 arch/arm/boot/dts/at91sam9x5ek.dtsi |    8 --------
 3 files changed, 22 insertions(+), 13 deletions(-)

diff --git a/arch/arm/boot/dts/at91sam9x5.dtsi b/arch/arm/boot/dts/at91sam9x5.dtsi
index af0419d..8264937 100644
--- a/arch/arm/boot/dts/at91sam9x5.dtsi
+++ b/arch/arm/boot/dts/at91sam9x5.dtsi
@@ -197,9 +197,9 @@
 				};
 
 				usart3 {
-					pinctrl_uart3: usart3-0 {
+					pinctrl_usart3: usart3-0 {
 						atmel,pins =
-							<2 23 0x2 0x1	/* PC22 periph B with pullup */
+							<2 22 0x2 0x1	/* PC22 periph B with pullup */
 							 2 23 0x2 0x0>;	/* PC23 periph B */
 					};
 
@@ -402,10 +402,12 @@
 				compatible = "atmel,at91sam9260-usart";
 				reg = <0xf801c000 0x200>;
 				interrupts = <5 4 5>;
-				atmel,use-dma-rx;
-				atmel,use-dma-tx;
 				pinctrl-names = "default";
-				pinctrl-0 = <&pinctrl_usart0>;
+				pinctrl-0 = <
+					&pinctrl_usart0
+					&pinctrl_usart0_rts
+					&pinctrl_usart0_cts
+				>;
 				status = "disabled";
 			};
 
diff --git a/arch/arm/boot/dts/at91sam9x5cm.dtsi b/arch/arm/boot/dts/at91sam9x5cm.dtsi
index 80ef0ba..a06acb7 100644
--- a/arch/arm/boot/dts/at91sam9x5cm.dtsi
+++ b/arch/arm/boot/dts/at91sam9x5cm.dtsi
@@ -84,6 +84,21 @@
 				reg = <0x7fa0000 0x60000>;
 			};
 		};
+
+
+		apb {
+			usart3: serial@f8028000 {
+				compatible = "atmel,at91sam9260-usart";
+				reg = <0xf8028000 0x200>;
+				interrupts = <8 4 5>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&pinctrl_usart3
+					&pinctrl_usart3_rts
+					&pinctrl_usart3_cts
+					>;
+				status = "okay";
+			};
+		};
 	};
 
 	gpio_keys {
diff --git a/arch/arm/boot/dts/at91sam9x5ek.dtsi b/arch/arm/boot/dts/at91sam9x5ek.dtsi
index b09e0c8..9a48c41 100644
--- a/arch/arm/boot/dts/at91sam9x5ek.dtsi
+++ b/arch/arm/boot/dts/at91sam9x5ek.dtsi
@@ -52,14 +52,6 @@
 				status = "okay";
 			};
 
-			usart1: serial@f8020000 {
-				status = "okay";
-			};
-
-			usart2: serial@f8024000 {
-				status = "okay";
-			};
-
 			macb0: ethernet@f802c000 {
 				phy-mode = "rmii";
 				status = "okay";
-- 
1.7.9.5

